enhancements/authentication/pod-security-admission-enforcement.md

```diff
@@ -0,0 +1,428 @@
+
+- With the `security.openshift.io/MinimallySufficientPodSecurityStandard` annotation, it can evaluate Namespaces that lack the `pod-security.kubernetes.io/enforce` label but have user-overridden warn or audit labels.
+- With `ValidatedSCCSubjectType`, the controller can classify issues arising from user-based SCC workloads separately.
+  Many of the remaining clusters with violations appear to involve workloads admitted by user SCCs.
+
+### Secure Rollout
+
+The Proposal section indicates that enforcement will be introduced first at the Namespace level and later at the global (cluster-wide) level.
+In addition to adjusting how the `OpenShiftPodSecurityAdmission` `FeatureGate` behaves, administrators need visibility and control throughout this transition.
+A new API is necessary to provide this flexibility.
+
+#### New API
+
+This API offers a gradual way to roll out Pod Security Admission enforcement to clusters.
+It gives users the ability to influence the rollout and see feedback on which Namespaces might violate Pod Security standards.
+
+```go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PSAEnforcementMode indicates the actual enforcement state of Pod Security Admission
+// in the cluster. Unlike PSATargetMode, which reflects the user’s desired or “target”
+// setting, PSAEnforcementMode describes the effective mode currently active.
+//
+// The modes define a progression from no enforcement, through label-based enforcement
+// to label-based with global config enforcement.
+type PSAEnforcementMode string
+
+const (
+	// EnforcementModePrivileged indicates that no Pod Security restrictions
+	// are effectively applied.
+	// This aligns with a pre-rollout or fully "privileged" cluster state,
+	// where neither enforce labels are set nor the global config enforces "Restricted".
+	EnforcementModePrivileged PSAEnforcementMode = "Privileged"
+
+	// EnfrocementModeLabel indicates that the cluster is enforcing Pod Security
+	// labels at the Namespace level (via the PodSecurityAdmissionLabelSynchronizationController),
+	// but the global kube-apiserver configuration is still "Privileged."
+	EnfrocementModeLabel PSAEnforcementMode = "LabelEnforcement"
+
+	// EnforcmentModeFull indicates that the cluster is enforcing
+	// labels at the Namespace level, and the global configuration has been set
+	// to "Restricted" on the kube-apiserver.
+	// This represents full enforcement, where both Namespace labels and the global config
+	// enforce Pod Security Admission restrictions.
+	EnforcmentModeFull PSAEnforcementMode = "FullEnforcement"
+)
+
+// PSATargetMode reflects the user’s chosen (“target”) enforcement level.
+type PSATargetMode string
+
+const (
```

Comment:
Is there any particular reason there isn't an option for the [Baseline](https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline) enforcement level? Is this enforcement level one that users may want to use over privileged or restricted?

Reply:
I mentioned this in the [Open Questions](https://github.com/openshift/enhancements/blob/f8f491154ab3a462566d11cfc15d29099162c2ad/enhancements/authentication/pod-security-admission-enforcement.md#baseline-clusters) section, that this would be a viable thing to do... I am not sure if our users would like to do that. They might, if they struggle to enforce restricted.

Reply:
Ah, I missed that open question. Thanks for the pointer to it. I can move further discussion there if you'd prefer.

Reply:
This is a good discussion to have. Lets move it there.

Reply:
https://github.com/openshift/enhancements/pull/1747/files#r1954607428